extends Node

# ğŸ‘‡ Signal is going to be emitted in two cases:
signal unlock_new_skin
# â˜ï¸ 1. If player doesn't own the skin and they have just purchased it.
# â˜ï¸ 2. If the player has purchased it in the past and we can learn that from querry.

var google_payment = null
# ğŸ‘‡ storing Product ID that you get from Play Console
var new_skin_sku = "new_player_skin"
# â˜ï¸ sku stands for "Stock Keeping Unit". This is an old naming scheme which Godot still uses.
# Google have changed it to using products.
# ğŸ‘‡ New skin token
var new_skin_token = ""


func _ready() -> void:
	# ğŸ‘‡ The GodotGooglePlayBilling PlugIn is a singleton
	# ğŸ‘‡ So we check if it is there or not
	if Engine.has_singleton("GodotGooglePlayBilling"):
		# ğŸ‘‡ Stores the name of global singleton that is returned.
		google_payment = Engine.get_singleton("GodotGooglePlayBilling")
		# google_payment.purchase("")
		MyUtility.add_log_msg("Android IAP support is enabled.")
		
		# ğŸ‘‡ Emitted after we receive a response from google
		google_payment.connected.connect(_on_connected)
		# ğŸ‘‡ Emitted in case if there is a connection error
		google_payment.connect_error.connect(_on_connect_error)
		# ğŸ‘‡ Emitted in case if connection is disconnected
		google_payment.disconnected.connect(_on_disconnected)
		# ğŸ‘‡ To make a call across the internet, we create this signal for success case
		google_payment.sku_details_query_completed.connect(_on_sku_details_query_completed)
		# ğŸ‘‡ To make a call across the internet, we create this signal for failure case
		google_payment.sku_details_querry_error.connect(_on_sku_details_query_error)
		
		google_payment.purchases_updated.connect(_on_purchases_updated)
		google_payment.purchases_error.connect(_on_purchase_error)
		# ğŸ‘‡ Creating signal to make a call to google to acknowledge the purchase
		google_payment.purchase_acknowledged.connect(_on_purchase_acknowledged) # ğŸ‘ˆ Purchase token (string)
		# ğŸ‘‡ Creating signal to hande any error encountered during purchase acknowledgement
		google_payment.purchase_acknowledgement_error.connect(_on_purchase_acknowledgement_error)
		# ğŸ‘‡ Creating signal to hande request the list of purchases from google
		google_payment.query_purchases_response.connect(_on_query_purchases_response)
		# ğŸ‘‡ FOR TESTING ONLY, to get rid of the purchase that we made in the past
		google_payment.payment.purchase_consumed.connect(_on_purchase_consumed)
		# ğŸ‘‡ FOR TESTING ONLY, to catch any error
		google_payment.payment.purchase_consumption_error.connect(_on_purchase_consumption_error)
		# ğŸ‘‡ Signal to start connection of app to google cloud service
		# ğŸ‘‡ Starts connection between us and google play.
		google_payment.startConnection()
	else:
		MyUtility.add_log_msg("Android IAP support is not available.")

func purchase_skin():
	if google_payment:
		var response = google_payment.purchase(new_skin_sku)
		MyUtility.add_log_msg("Purchase attempted, response " + str(response.status))
		# ğŸ‘‡ This is the local response status whetherâ˜ï¸above function call succeded OR not.
		if response.status != OK:
			MyUtility.add_log_msg("Error purchasing skin!")

# ğŸ‘‡ FOR TESTING ONLY
func reset_purchases():
	if google_payment:
		if !new_skin_token.is_empty():
			# ğŸ‘‡ It takes in the "new_skin_token" and it gets rid of the purchase that we made in the past.
			# ğŸ‘‡ So that we can buy again for testing and see if it can be bought again and test other end cases.
			# ğŸ‘‡ Make a call to google and tell that we want to consume this in-app product. We don't want to have it anymore.
			google_payment.consumePurchase(new_skin_token)
			# â˜ï¸ And Google is gonna do that inside their servers and send a response. For which we will create signals above.

func _on_connected():
	MyUtility.add_log_msg("Connected")
	# ğŸ‘‡ Takes 2 input fields: [an array of strings], "string"
	# ğŸ‘‰ Two types of SKUs in google: 1. In app-purchase or in-app products, 2. Subscription
	# ğŸ‘‰ Because you can have more than one products to sell, just add sku name in the array: [sku1, sku2]
	# ğŸ‘‡ This will make a call across the internet to google to ask details about "new_skin_sku".
	# ğŸ‘‡ For which we will make 2 signals in the ready() function
	google_payment.querrySkuDetails([new_skin_sku], "inapp") # ğŸ‘ˆ For subscription, it will use string "subs"

##### ğŸ‘‡ğŸ‘‡ğŸ‘‡ Always Look at the documentation to see what kind of arguements they accept ğŸ‘‡ğŸ‘‡ğŸ‘‡

func _on_connect_error(response_id, debug_msg):
	MyUtility.add_log_msg("Connect error, response id: " + str(response_id) + " debug msg: " + debug_msg)

func _on_disconnected():
	MyUtility.add_log_msg("Disconnected")
# ğŸ‘‡ Ask google at the start of the game about the previous
# ğŸ‘‡ purchases we made.
func _on_sku_details_query_completed(skus):
	MyUtility.add_log_msg("Sku details query completed")
	# ğŸ‘‡ Since there can be multiple products, we create a loop
	for sku in skus:
		MyUtility.add_log_msg("Sku:")
		MyUtility.add_log_msg(str(sku))
		# â˜ï¸ Each one of these SKUs that we have in the array are going to be a dictionary
		# â˜ï¸ and it is going to contain information about the in-app product.
	# ğŸ‘‡ "querryPurchases" is going to ask google what are the purchases 
	# ğŸ‘‡ that have been made? Googlw will share a list of all the prchases that have been made.
	google_payment.querryPurchases("inapp")

func _on_sku_details_query_error(response_id, error_message, skus):
	MyUtility.add_log_msg("Sku querry error, response id: " + str(response_id) + 
	", message: " + str(error_message) + ", skus: " + str(skus))

func _on_purchases_updated(purchases):
	# â˜ï¸ For now, this array contains a single item that can be purchased. Hence, no need of "for" loop.
	# ğŸ‘‰ In case you are expecting multiple purchases, you should definitely create a loop for these.
	if purchases.size() > 0: # ğŸ‘ˆ Checking the size of the array to know if there is an item present OR not
		var purchase = purchases[0]
		#  This â˜ ï¸is a dictionary.It will contain information that has to do with the purchases.
		var purchase_sku = purchase["skus"][0] # ğŸ‘ˆ The "skus" here is an array filled with "skus".
		# â˜ï¸This here should be equal to var new_skin_sku. We try to get its first element.
		# â˜ï¸ï¸We need to check if it is equal to the one we have or not.
		# â˜ï¸If you have multiple in-app products, you should definitely check this.
		MyUtility.add_log_msg("Purchased item with sku: " + purchase_sku)
		if purchase_sku == new_skin_sku:
			new_skin_token = purchase.purchase_token # ğŸ‘ˆ This token is generated by google and it signifies all purchase. It represents the purchase. It is unique to us.
			# ğŸ‘‡ Now we need to acknowledge the purchase. If we don't, google will refund player's money
			google_payment.acknowledgePurchase(purchase.purchase_token)
		
func _on_purchase_error(response_id, error_message):
	MyUtility.add_log_msg("Purchase error, response id: " 
	+ str(response_id) + " error msg: " + error_message)

func _on_purchase_acknowledged(purchase_token):
	MyUtility.add_log_msg("Purchase acknowledged successfully!")
	# ğŸ‘‡ Works in this case because there is only single in-app product
	if !new_skin_token.is_empty():
		# ğŸ‘‡ new_skin_token stored in line 100 == purchase_token sent by google
		if new_skin_token == purchase_token:
			MyUtility.add_log_msg("Unlocking new skin.")
			# ğŸ‘‡ Called when item is purchased for the first time
			unlock_new_skin.emit()
		# ğŸ‘‰ If there are more than one product purchases, 
		# ğŸ‘‰ you will create similar "if" cases for them as aboveâ˜ï¸

func _on_purchase_acknowledgement_error(response_id, error_message, purchase_token):
	MyUtility.add_log_msg("Purchase acknowledgement error, response id: " + str(response_id) + 
	", message: " + str(error_message) + ", token: " + str(purchase_token))

func _on_query_purchases_response(querry_result):# ğŸ‘ˆ "querry_result" should have an array of dictionar. It itself is not an array of dictionary.
	# ğŸ‘‡ The querry_result will have a status.This will be checked first.
	if querry_result.status == OK:
		MyUtility.add_log_msg("QUerry purchases was successful.")
		# ğŸ‘‡ Proceeding to get the array of dictionary from this result
		var purchases = querry_result.purchases 
		# â˜ï¸ "purchases" is an array filled with dictionaries that have 
		# â˜ï¸ information about each purchase we made.
		MyUtility.add_log_msg("purchases") # ğŸ‘ˆ Printing it to just ahve a look at it.
		var purchase = purchases[0]
		# â˜ï¸ this will have a key called "skus". 
		var purchase_sku = purchase["skus"][0] # ğŸ‘ˆ purchase dictionar is going to be the same
		if new_skin_sku == purchase_sku:
			# ğŸ‘‡ Getting token outside of the ourchase and saving it here
			new_skin_token = purchase.purchase_token
			# ğŸ‘‡ We are going to check if this purchase has been acknowledged OR not
			if !purchase.is_acknowledges:
				google_payment.acknowledgePurchase(purchase.purchase_token)
			else:
				unlock_new_skin.emit()
				MyUtility.add_log_msg("Unlocking new skin because it was purchased previously.")
	else:
		MyUtility.add_log_msg("Querry purchases failed")

# ğŸ‘‡ FOR TESTING ONLY
func _on_purchase_consumed(purchase_token):
	MyUtility.add_log_msg("Purchases consumed successfully!")

# ğŸ‘‡ FOR TESTING ONLY
func _on_purchase_consumption_error(response_id, error_message, purchase_token):
	MyUtility.add_log_msg("Purchase consumption acknowledgement error, response id: " + str(response_id) + 
	", message: " + str(error_message) + ", token: " + str(purchase_token))
